File: frontend/types/attendance.ts
// Per-class breakdown row (from /attendance/school → by_class)
export interface AttendanceReportItem {
  class_id: string;
  class_name: string;
  total_students: number;
  present: number;
  absent: number;
  late: number;
  excused: number;
  attendance_rate: number;
}

// School-wide summary (from GET /attendance/school)
export interface SchoolAttendanceSummary {
  date: string;
  total_students: number;
  present: number;
  absent: number;
  late: number;
  excused: number;
  attendance_rate: number;
  by_class: AttendanceReportItem[];
}

// Class-level summary (from GET /attendance/class/{class_id}/summary)
export interface ClassAttendanceSummary {
  date?: string;
  class_id?: string;
  class_name?: string;
  total_students: number;
  present: number;
  absent: number;
  late: number;
  excused?: number;
  attendance_rate?: number;
}

// Individual attendance record (from GET /attendance/class/{class_id})
export interface AttendanceRecord {
  id: string;
  student_id: string;
  student_name: string;
  class_id: string;
  date: string;
  status: 'present' | 'absent' | 'late' | 'excused';
  note?: string;
  marked_by_name: string;
}

// Mark attendance request
export interface MarkAttendanceEntry {
  student_id: string;
  status: 'present' | 'absent' | 'late' | 'excused';
  note?: string;
}

export interface MarkAttendanceRequest {
  class_id: string;
  date: string;
  records: MarkAttendanceEntry[];
}

// Mark attendance response
export interface MarkAttendanceResponse {
  marked_count: number;
  date: string;
  class_name: string;
  already_marked: boolean;
}

// Check attendance response
export interface CheckAttendanceResponse {
  is_marked: boolean;
  date: string;
  marked_at?: string;
  marked_by?: string;
}

export interface AttendanceReportParams {
  date?: string;
  class_id?: string;
}

File: backend/app/schemas/attendance.py
import uuid
from datetime import date, datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class AttendanceStatusEnum(str, Enum):
    present = "present"
    absent = "absent"
    late = "late"
    excused = "excused"


# ---------------------------------------------------------------------------
# Mark attendance (request)
# ---------------------------------------------------------------------------

class AttendanceRecord(BaseModel):
    student_id: uuid.UUID
    status: AttendanceStatusEnum
    note: Optional[str] = Field(None, max_length=500)


class MarkAttendanceRequest(BaseModel):
    class_id: uuid.UUID
    date: date
    records: list[AttendanceRecord] = Field(..., min_length=1)


# ---------------------------------------------------------------------------
# Responses
# ---------------------------------------------------------------------------

class AttendanceResponse(BaseModel):
    id: uuid.UUID
    student_id: uuid.UUID
    student_name: str
    class_id: uuid.UUID
    date: date
    status: AttendanceStatusEnum
    note: Optional[str]
    marked_by_name: str

    model_config = {"from_attributes": True}


class MarkAttendanceResponse(BaseModel):
    marked_count: int
    date: date
    class_name: str
    already_marked: bool


class CheckAttendanceResponse(BaseModel):
    is_marked: bool
    date: date
    marked_at: Optional[datetime]
    marked_by: Optional[str]


# ---------------------------------------------------------------------------
# Summary schemas
# ---------------------------------------------------------------------------

class AttendanceUpdate(BaseModel):
    status: Optional[AttendanceStatusEnum] = None
    note: Optional[str] = Field(None, max_length=500)


class AttendanceSummary(BaseModel):
    date: Optional[date]
    class_id: uuid.UUID
    class_name: str
    total_students: int
    present: int
    absent: int
    late: int
    excused: int
    attendance_rate: float


class SchoolAttendanceSummary(BaseModel):
    date: date
    total_students: int
    present: int
    absent: int
    late: int
    excused: int
    attendance_rate: float
    by_class: list[AttendanceSummary]


class StudentAttendanceSummary(BaseModel):
    student_name: str
    total_days: int
    present: int
    absent: int
    late: int
    excused: int
    attendance_rate: float
